### PR TITLE
Implemented TransferSaga.

### PR DESCRIPTION
**Pull request by polonzer**
https://lykkex.atlassian.net/browse/LWDEV-5664

### CODE REVIEW

File: src/Lykke.Job.TransactionHandler.AzureRepositories/Offchain/OffchainRequestRepository.cs [hunk 1]
```diff
@@ -117,45 +115,12 @@ public OffchainRequestRepository(INoSQLTableStorage<OffchainRequestEntity> table
             var id = Guid.NewGuid().ToString();
 
             var byClient = OffchainRequestEntity.ByClient.Create(id, transferId, clientId, assetId, type, transferType, serverLock);
-            await _table.InsertAsync(byClient);
+            await _table.InsertOrMergeAsync(byClient);
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Job.TransactionHandler.AzureRepositories/Offchain/OffchainRequestRepository.cs]: InsertOrReplace is more suitable

File: src/Lykke.Job.TransactionHandler.AzureRepositories/Offchain/OffchainRequestRepository.cs [hunk 2]
```diff
@@ -117,45 +115,12 @@ public OffchainRequestRepository(INoSQLTableStorage<OffchainRequestEntity> table
...
 
             var byRecord = OffchainRequestEntity.ByRecord.Create(id, transferId, clientId, assetId, type, transferType, serverLock);
-            await _table.InsertAsync(byRecord);
+            await _table.InsertOrMergeAsync(byRecord);
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Job.TransactionHandler.AzureRepositories/Offchain/OffchainRequestRepository.cs]: InsertOrReplace is more suitable

File: src/Lykke.Job.TransactionHandler.AzureRepositories/Offchain/OffchainTransferRepository.cs [hunk 3]
```diff
@@ -64,7 +64,7 @@ public OffchainTransferRepository(INoSQLTableStorage<OffchainTransferEntity> sto
         {
             var entity = OffchainTransferEntity.ByCommon.Create(transactionId, clientId, assetId, amount, type, externalTransferId, orderId, channelClosing);
 
-            await _storage.InsertAsync(entity);
+            await _storage.InsertOrMergeAsync(entity);
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Job.TransactionHandler.AzureRepositories/Offchain/OffchainTransferRepository.cs]: InsertOrReplace is more suitable

File: src/Lykke.Job.TransactionHandler/Handlers/EthereumCommandHandler.cs [hunk 4]
```diff
@@ -21,26 +26,35 @@ public class EthereumCommandHandler
...
+            _transactionService = transactionService ?? throw new ArgumentNullException(nameof(transactionService));
             _settings = settings ?? throw new ArgumentNullException(nameof(settings));
             _retryTimeout = retryTimeout;
         }
 
-        public async Task<CommandHandlingResult> Handle(Commands.ProcessEthereumCashoutCommand command)
+        public async Task<CommandHandlingResult> Handle(ProcessEthereumCashoutCommand command)
         {
-            await _log.WriteInfoAsync(nameof(EthereumCommandHandler), nameof(Commands.ProcessEthereumCashoutCommand), command.ToJson(), "");
+            await _log.WriteInfoAsync(nameof(EthereumCommandHandler), nameof(ProcessEthereumCashoutCommand), command.ToJson(), "");
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Job.TransactionHandler/Handlers/EthereumCommandHandler.cs]: You can do ```CreateComponentScope``` in the ctor and omit ```nameof(EthereumCommandHandler)``` in every log call then

File: src/Lykke.Job.TransactionHandler/Handlers/EthereumCommandHandler.cs [hunk 5]
```diff
@@ -79,11 +93,72 @@ public class EthereumCommandHandler
             if (errorMessage != null)
             {
                 await _ethClientEventLogs.WriteEvent(command.ClientId, Event.Error, new { Request = command.TransactionId, Error = errorMessage }.ToJson());
-                await _log.WriteErrorAsync(nameof(EthereumCommandHandler), nameof(Commands.ProcessEthereumCashoutCommand), command.ToJson(), new Exception(errorMessage));
+                await _log.WriteErrorAsync(nameof(EthereumCommandHandler), nameof(ProcessEthereumCashoutCommand), command.ToJson(), new Exception(errorMessage));
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Job.TransactionHandler/Handlers/EthereumCommandHandler.cs]: You an use sync version of method, where context could be an object, so you can avoid ```ToJson()``` call

File: src/Lykke.Job.TransactionHandler/Projections/OperationHistoryProjection.cs [hunk 6]
```diff
@@ -55,8 +58,71 @@ public class OperationHistoryProjection
...
+            var amountNoFee = evt.AmountNoFee;
+
+            var context = await _transactionService.GetTransactionContext<TransferContextData>(transactionId);
+
+            //Get eth request if it is ETH transfer
+            var ethTxRequest = await _ethereumTransactionRequestRepository.GetAsync(Guid.Parse(transactionId));
+
+            //Get client wallets
+            var destWallet = await _walletCredentialsRepository.GetAsync(message.ToClientid);
+            var sourceWallet = await _walletCredentialsRepository.GetAsync(message.FromClientId);
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Job.TransactionHandler/Projections/OperationHistoryProjection.cs]: Will be good to add request with clientId collections in the repository for this case, to request client wallets with single round trip to the table storage

File: src/Lykke.Job.TransactionHandler.AzureRepositories/BitCoin/WalletCredentialsRepository.cs [hunk 7]
```diff
@@ -60,5 +62,16 @@ public WalletCredentialsRepository(INoSQLTableStorage<WalletCredentialsEntity> t
...
         }
+
+        public async Task<IEnumerable<IWalletCredentials>> GetWalletsAsync(IEnumerable<string> clientIds)
+        {
+            var partitionKey = WalletCredentialsEntity.ByClientId.GeneratePartitionKey();
+
+            return await _tableStorage.GetDataAsync(
+                partitionKey,
+                clientIds.Select(WalletCredentialsEntity.ByClientId.GenerateRowKey),
+                100,
```
Comment (MikhailGerasimov) [inline src/Lykke.Job.TransactionHandler.AzureRepositories/BitCoin/WalletCredentialsRepository.cs]: A magic number

Comment (polonzer) [reply src/Lykke.Job.TransactionHandler.AzureRepositories/BitCoin/WalletCredentialsRepository.cs]: it's a default value for this method.

Comment (MikhailGerasimov) [reply src/Lykke.Job.TransactionHandler.AzureRepositories/BitCoin/WalletCredentialsRepository.cs]: It is really not obvious when I do review in browser.

Comment (polonzer) [reply src/Lykke.Job.TransactionHandler.AzureRepositories/BitCoin/WalletCredentialsRepository.cs]: Agreed. Fixed.

**Pull request merged**